[PyTorch Edge] share tensors in mobile with new api

cccclai opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #57888 [PyTorch Edge] bytecode version bump to v5 and enable share constant table
* **#58182 [PyTorch Edge] share tensors in mobile with new api**

As title, the v5 model format will be
```
(base) chenlai@chenlai-mp reuse_constant % zipinfo /Users/chenlai/Documents/pytorch/reuse_constant/tmp/zip/script_module_v5_unify.ptl
Archive:  /Users/chenlai/Documents/pytorch/reuse_constant/tmp/zip/script_module_v5_unify.ptl
Zip file size: 3120 bytes, number of entries: 7
-rw----     0.0 fat       77 bl stor 80-000-00 00:00 script_module_v4_unify/data.pkl
-rw----     0.0 fat      240 bl defN 80-000-00 00:00 script_module_v4_unify/code/__torch__/___torch_mangle_5.py
-rw----     0.0 fat      422 bl defN 80-000-00 00:00 script_module_v4_unify/code/__torch__/___torch_mangle_5.py.debug_pkl
-rw----     0.0 fat       64 bl stor 80-000-00 00:00 script_module_v4_unify/constants/140245072983168.storage
-rw----     0.0 fat      172 bl stor 80-000-00 00:00 script_module_v4_unify/constants.pkl
-rw----     0.0 fat      678 bl stor 80-000-00 00:00 script_module_v4_unify/bytecode.pkl
-rw----     0.0 fat        2 bl stor 80-000-00 00:00 script_module_v4_unify/version
7 files, 1655 bytes uncompressed, 1453 bytes compressed:  12.2%
```
bytecode.pkl is:
```
(5,
 ('__torch__.___torch_mangle_5.TestModule.forward',
  (('instructions',
    (('STOREN', 1, 2),
     ('DROPR', 1, 0),
     ('LOADC', 0, 0),
     ('LOADC', 1, 0),
     ('MOVE', 2, 0),
     ('OP', 0, 0),
     ('LOADC', 1, 0),
     ('OP', 1, 0),
     ('RET', 0, 0))),
   ('operators', (('aten::add', 'int'), ('aten::add', 'Scalar'))),
   ('constants',
    (torch._utils._rebuild_tensor_v2(pers.obj(('storage',
          torch.DoubleStorage,
          '140245072983168.storage',
          'cpu',
          8),),
       0,
       (2, 4),
       (4, 1),
       False,
       collections.OrderedDict()),
     1)),
   ('types', ()),
   ('register_size', 2)),
  (('arguments',
    ((('name', 'self'),
      ('type', '__torch__.___torch_mangle_5.TestModule'),
      ('default_value', None)),
     (('name', 'y'), ('type', 'int'), ('default_value', None)))),
   ('returns',
    ((('name', ''), ('type', 'Tensor'), ('default_value', None)),)))))
```

constants.pkl is:
```
(torch._utils._rebuild_tensor_v2(pers.obj(('storage', torch.DoubleStorage, '140245072983168.storage', 'cpu', 8),),
   0,
   (2, 4),
   (4, 1),
   False,
   collections.OrderedDict()),)
```

Both tensors will refer to the tensor in at the path `script_module_v4_unify/constants/140245072983168.storage`.

## Note
According to unify format, all tensors will be written to the folder `.data`, however, torch.jit.load() can't handle the unified format at this moment, so this change will write tensors at the `constants` folders, and mobile will write/read tensors from `constants` folder. such that the model can be interpreted by both jit and mobile.

Differential Revision: [D28375257](https://our.internmc.facebook.com/intern/diff/D28375257/)

Commit:
[PyTorch Edge] share tensors in mobile with new api

As title, the v5 model format will be
```
(base) chenlai@chenlai-mp reuse_constant % zipinfo /Users/chenlai/Documents/pytorch/reuse_constant/tmp/zip/script_module_v5_unify.ptl
Archive:  /Users/chenlai/Documents/pytorch/reuse_constant/tmp/zip/script_module_v5_unify.ptl
Zip file size: 3120 bytes, number of entries: 7
-rw----     0.0 fat       77 bl stor 80-000-00 00:00 script_module_v4_unify/data.pkl
-rw----     0.0 fat      240 bl defN 80-000-00 00:00 script_module_v4_unify/code/__torch__/___torch_mangle_5.py
-rw----     0.0 fat      422 bl defN 80-000-00 00:00 script_module_v4_unify/code/__torch__/___torch_mangle_5.py.debug_pkl
-rw----     0.0 fat       64 bl stor 80-000-00 00:00 script_module_v4_unify/constants/140245072983168.storage
-rw----     0.0 fat      172 bl stor 80-000-00 00:00 script_module_v4_unify/constants.pkl
-rw----     0.0 fat      678 bl stor 80-000-00 00:00 script_module_v4_unify/bytecode.pkl
-rw----     0.0 fat        2 bl stor 80-000-00 00:00 script_module_v4_unify/version
7 files, 1655 bytes uncompressed, 1453 bytes compressed:  12.2%
```
bytecode.pkl is:
```
(5,
 ('__torch__.___torch_mangle_5.TestModule.forward',
  (('instructions',
    (('STOREN', 1, 2),
     ('DROPR', 1, 0),
     ('LOADC', 0, 0),
     ('LOADC', 1, 0),
     ('MOVE', 2, 0),
     ('OP', 0, 0),
     ('LOADC', 1, 0),
     ('OP', 1, 0),
     ('RET', 0, 0))),
   ('operators', (('aten::add', 'int'), ('aten::add', 'Scalar'))),
   ('constants',
    (torch._utils._rebuild_tensor_v2(pers.obj(('storage',
          torch.DoubleStorage,
          '140245072983168.storage',
          'cpu',
          8),),
       0,
       (2, 4),
       (4, 1),
       False,
       collections.OrderedDict()),
     1)),
   ('types', ()),
   ('register_size', 2)),
  (('arguments',
    ((('name', 'self'),
      ('type', '__torch__.___torch_mangle_5.TestModule'),
      ('default_value', None)),
     (('name', 'y'), ('type', 'int'), ('default_value', None)))),
   ('returns',
    ((('name', ''), ('type', 'Tensor'), ('default_value', None)),)))))
```

constants.pkl is:
```
(torch._utils._rebuild_tensor_v2(pers.obj(('storage', torch.DoubleStorage, '140245072983168.storage', 'cpu', 8),),
   0,
   (2, 4),
   (4, 1),
   False,
   collections.OrderedDict()),)
```

Both tensors will refer to the tensor in at the path `script_module_v4_unify/constants/140245072983168.storage`.

## Note
According to unify format, all tensors will be written to the folder `.data`, however, torch.jit.load() can't handle the unified format at this moment, so this change will write tensors at the `constants` folders, and mobile will write/read tensors from `constants` folder. such that the model can be interpreted by both jit and mobile.

Differential Revision: [D28375257](https://our.internmc.facebook.com/intern/diff/D28375257/)

[ghstack-poisoned]

raziel commented on torch/csrc/jit/mobile/backport_manager.cpp:
```diff
@@ -36,16 +37,23 @@ bool update_bytecode_version(
-        record.find(kArchiveNameBytecode) == std::string::npos) {
+    bool skip = false;
+    for (const auto& excluded_file : excluded_files) {
+      if (record.find(excluded_file) != std::string::npos) {
```
is substring safe here?
otherwise, should be be strictly checking for the full path (and thus use the set's find function instead)?

iseeyuan replied:
Good point. A record includes the path but expluded_file does not. Is it possible that there are two files with the same name but in different paths?

cccclai replied:
Yeah it's a good point. Given our current model schema, it should be fine. But I agree full path will be safer. I add a bit more context here and more comment in the source code and hope it's easier to understand. `reader.getAllRecords()` will return a file list, say:
```
script_module_v4_unify/data.pkl
script_module_v4_unify/code/__torch__/___torch_mangle_5.py
script_module_v4_unify/code/__torch__/___torch_mangle_5.py.debug_pkl
script_module_v4_unify/constants/0
script_module_v4_unify/constants.pkl
script_module_v4_unify/bytecode/0
script_module_v4_unify/bytecode.pkl
script_module_v4_unify/version
```
If we want to skip writing bytecode archive, we need to skip both `script_module_v4_unify/bytecode/0` (the actual tensor data) and `script_module_v4_unify/bytecode.pkl`. We currently don't have a way to know how many tensors in total without reading `bytecode.pkl` + `bytecode` folder together. I'm looking at an easier way to load pickle file, since it's just a zip folder and it shouldn't need any pytorch related knowledge. After it's done, we can leverage it here. It can be part of the work in T90180710. Let me add it in the comment so we can keep track of it.

cccclai replied:
Use regex to match instead in the latest commit.

cccclai replied:
Ok internal CI doesn't like regex. Switch back with more strict substring searching

raziel commented on torch/csrc/jit/mobile/import.h:
```diff
@@ -85,6 +89,8 @@ c10::intrusive_ptr<c10::ivalue::Object> objLoaderMobile(
     at::StrongTypePtr type,
     at::IValue input,
     std::shared_ptr<mobile::CompilationUnit> mobile_compilation_unit);
+bool tensorInConstantsArchive(
```
maybe IsTensorIn...

some docs would be nice too.

cccclai replied:
Rename to `isTensorInBytecodeArchive` in latest version. As a note, it is *not* `isTensorInConstantArchive`. I realize that there are always tensors in `constants` archive, but not `bytecode` archive in the future model, so I change the logic.

cccclai replied:
In terms of why unit test didn't catch it? Likely all tensors in existing testing models from `bytecode` are exactly the same as `constants`...anyway, it's better to use `isTensorInBytecodeArchive` logic.

raziel reviewed (commented):
Thanks!

I suggest focusing the description on the goal of the changes rather than on whether this is the new or old api.

I'm not familiar with the changes in serialization so @Lilyjjo and @suo should take a closer look (plus @iseeyuan for the mobile specific details).

iseeyuan commented on caffe2/serialize/versions.h:
```diff
@@ -67,7 +67,7 @@ constexpr uint64_t kProducedFileFormatVersion = 0x3L;
 //  0x4L: (Comment missing)
 //  0x4L: (update) Added schema to function tuple. Forward-compatible change.
-constexpr uint64_t kProducedBytecodeVersion = 0x4L;
+constexpr uint64_t kProducedBytecodeVersion = 0x5L;
```
Are we bumping the version in the same PR?

iseeyuan replied:
Chatted offline. It's for testing purpose. Please remember to change it back to 4 before landing it.

cccclai replied:
Ideally yeah, consider the change is quite minimum and all the test in place.

cccclai replied:
Move the version bump to #57888

iseeyuan commented on torch/csrc/jit/serialization/export_module.cpp:
```diff
@@ -346,16 +346,23 @@ void ScriptModuleSerializer::serialize(
+        /*archive_name=*/kArchiveNameConstants,
+        /*archive_dir=*/"",
+        /*tensor_dir=*/"constants/",
+        true);
```
Add comment on this arg, like the args above.

cccclai replied:
Add it in the latest commit.

iseeyuan commented on torch/csrc/jit/serialization/pickler.cpp:
```diff
@@ -310,6 +310,8 @@ void Pickler::pushStorageOfTensor(const at::Tensor& tensor) {
   if (get_tensor_id_ != nullptr) {
     root_key = get_tensor_id_(tensor);
+  } else {
+    root_key = c10::to_string(tensor_data_.size());
```
Looks a good change to me. @suo and @Lilyjjo please confirm if it's good.

cccclai replied:
It's a temporary fix and @Lilyjjo addressed it in the stack

cccclai replied:
Rebase and there is no change in `pickler.cpp` now.

iseeyuan reviewed (commented):
I think it's good to share the API in general. Would the tests be on top of this PR?

Commit:
Update on "[PyTorch Edge] share tensors in mobile with new api"


As title, the v5 model format will be
```
(base) chenlai@chenlai-mp reuse_constant % zipinfo /Users/chenlai/Documents/pytorch/reuse_constant/tmp/zip/script_module_v5_unify.ptl
Archive:  /Users/chenlai/Documents/pytorch/reuse_constant/tmp/zip/script_module_v5_unify.ptl
Zip file size: 3120 bytes, number of entries: 7
-rw----     0.0 fat       77 bl stor 80-000-00 00:00 script_module_v4_unify/data.pkl
-rw----     0.0 fat      240 bl defN 80-000-00 00:00 script_module_v4_unify/code/__torch__/___torch_mangle_5.py
-rw----     0.0 fat      422 bl defN 80-000-00 00:00 script_module_v4_unify/code/__torch__/___torch_mangle_5.py.debug_pkl
-rw----     0.0 fat       64 bl stor 80-000-00 00:00 script_module_v4_unify/constants/140245072983168.storage
-rw----     0.0 fat      172 bl stor 80-000-00 00:00 script_module_v4_unify/constants.pkl
-rw----     0.0 fat      678 bl stor 80-000-00 00:00 script_module_v4_unify/bytecode.pkl
-rw----     0.0 fat        2 bl stor 80-000-00 00:00 script_module_v4_unify/version
7 files, 1655 bytes uncompressed, 1453 bytes compressed:  12.2%
```
bytecode.pkl is:
```
(5,
 ('__torch__.___torch_mangle_5.TestModule.forward',
  (('instructions',
    (('STOREN', 1, 2),
     ('DROPR', 1, 0),
     ('LOADC', 0, 0),
     ('LOADC', 1, 0),
     ('MOVE', 2, 0),
     ('OP', 0, 0),
     ('LOADC', 1, 0),
     ('OP', 1, 0),
     ('RET', 0, 0))),
   ('operators', (('aten::add', 'int'), ('aten::add', 'Scalar'))),
   ('constants',
    (torch._utils._rebuild_tensor_v2(pers.obj(('storage',
          torch.DoubleStorage,
          '140245072983168.storage',
          'cpu',
          8),),
       0,
       (2, 4),
       (4, 1),
       False,
       collections.OrderedDict()),
     1)),
   ('types', ()),
   ('register_size', 2)),
  (('arguments',
    ((('name', 'self'),
      ('type', '__torch__.___torch_mangle_5.TestModule'),
      ('default_value', None)),
     (('name', 'y'), ('type', 'int'), ('default_value', None)))),
   ('returns',
    ((('name', ''), ('type', 'Tensor'), ('default_value', None)),)))))
```

constants.pkl is:
```
(torch._utils._rebuild_tensor_v2(pers.obj(('storage', torch.DoubleStorage, '140245072983168.storage', 'cpu', 8),),
   0,
   (2, 4),
   (4, 1),
   False,
   collections.OrderedDict()),)
```

Both tensors will refer to the tensor in at the path `script_module_v4_unify/constants/140245072983168.storage`.

## Note
According to unify format, all tensors will be written to the folder `.data`, however, torch.jit.load() can't handle the unified format at this moment, so this change will write tensors at the `constants` folders, and mobile will write/read tensors from `constants` folder. such that the model can be interpreted by both jit and mobile.

Differential Revision: [D28375257](https://our.internmc.facebook.com/intern/diff/D28375257/)

[ghstack-poisoned]

cccclai commented on torch/csrc/jit/mobile/backport_manager.cpp:
```diff
@@ -36,16 +37,23 @@ bool update_bytecode_version(
   return false;
 }
 
-void copy_non_bytecode(
```
Added in the latest commit.

Lilyjjo commented on torch/csrc/jit/serialization/export_module.cpp:
```diff
@@ -343,16 +343,23 @@ void ScriptModuleSerializer::serialize(
+        /*archive_name=*/"constants",
+        /*archive_dir=*/"",
+        /*tensor_dir=*/"constants/",
+        /*tensor_cdata_naming_scheme=*/true);
```
Just a note, this change will introduce jitter between saves & loads of the same model since now the tensor names are no longer always the same 0/1/2/3/4 and instead are based off of the tensor's storage's c-memory pointer

Lilyjjo replied:
@cccclai is there a reason why you didn't also set the `tensor_cdata_naming_scheme` to true for the writeArchive call on `:334`? I might be misunderstanding the flow of mobile serialization, but I know that tensors can be written out to a `data/` folder instead of the `constants/` folder and you'd want those also to use the `tensor_cdata_naming_scheme` and `constants/` location for those tensors also

suo replied:
Ah this is a format thing. We actually write tensors to two locations today: the data folder, which contains tensors that are explicitly part of the module state (like parameters and buffers), and the constants folder, which contains tensors that are `prim::Constants` in the graph. 

As far as I understand the major goal of this work is to share the constants between the bytecode and TS code, since the data folder is already shared. Is this right @cccclai?

cccclai replied:
Good question! At this moment, the majority (probably all, I'm just being conservative) of duplicated tensors are between `constants` and `bytecode`. Also, for mobile, there is an `optimize_for_mobile()` step to convert tensors from parameters (in `data`) to `constants`, and ideally, there shouldn't be any duplicated tensors between `data` and `bytecode` after `optimized_for_mobile()` step. Given that it's unlikely to have duplication, I think we can wait until the unify format

cccclai replied:
@suo yeah this statement is correct.
> the major goal of this work is to share the constants between the bytecode and TS code, since the data folder is already shared

iseeyuan replied:
@suo Yes, we shared data (parameters) but not constants between jit and bytecode. For this PR we also share constants.

Lilyjjo replied:
Awesome, this makes sense then! :) 🚢 🚢

suo reviewed (commented):
I only looked at the `export_module.cpp` changes but they seem small and self-contained, lgtm. I think the smallness of the changes is a good sign we are on the right track with regard to unification!

Will leave it to @raziel or @iseeyuan to accept

Commit:
Update on "[PyTorch Edge] share tensors in mobile with new api"


As title, the v5 model format will be
```
(base) chenlai@chenlai-mp reuse_constant % zipinfo /Users/chenlai/Documents/pytorch/reuse_constant/tmp/zip/script_module_v5_unify.ptl
Archive:  /Users/chenlai/Documents/pytorch/reuse_constant/tmp/zip/script_module_v5_unify.ptl
Zip file size: 3120 bytes, number of entries: 7
-rw----     0.0 fat       77 bl stor 80-000-00 00:00 script_module_v4_unify/data.pkl
-rw----     0.0 fat      240 bl defN 80-000-00 00:00 script_module_v4_unify/code/__torch__/___torch_mangle_5.py
-rw----     0.0 fat      422 bl defN 80-000-00 00:00 script_module_v4_unify/code/__torch__/___torch_mangle_5.py.debug_pkl
-rw----     0.0 fat       64 bl stor 80-000-00 00:00 script_module_v4_unify/constants/140245072983168.storage
-rw----     0.0 fat      172 bl stor 80-000-00 00:00 script_module_v4_unify/constants.pkl
-rw----     0.0 fat      678 bl stor 80-000-00 00:00 script_module_v4_unify/bytecode.pkl
-rw----     0.0 fat        2 bl stor 80-000-00 00:00 script_module_v4_unify/version
7 files, 1655 bytes uncompressed, 1453 bytes compressed:  12.2%
```
bytecode.pkl is:
```
(5,
 ('__torch__.___torch_mangle_5.TestModule.forward',
  (('instructions',
    (('STOREN', 1, 2),
     ('DROPR', 1, 0),
     ('LOADC', 0, 0),
     ('LOADC', 1, 0),
     ('MOVE', 2, 0),
     ('OP', 0, 0),
     ('LOADC', 1, 0),
     ('OP', 1, 0),
     ('RET', 0, 0))),
   ('operators', (('aten::add', 'int'), ('aten::add', 'Scalar'))),
   ('constants',
    (torch._utils._rebuild_tensor_v2(pers.obj(('storage',
          torch.DoubleStorage,
          '140245072983168.storage',
          'cpu',
          8),),
       0,
       (2, 4),
       (4, 1),
       False,
       collections.OrderedDict()),
     1)),
   ('types', ()),
   ('register_size', 2)),
  (('arguments',
    ((('name', 'self'),
      ('type', '__torch__.___torch_mangle_5.TestModule'),
      ('default_value', None)),
     (('name', 'y'), ('type', 'int'), ('default_value', None)))),
   ('returns',
    ((('name', ''), ('type', 'Tensor'), ('default_value', None)),)))))
```

constants.pkl is:
```
(torch._utils._rebuild_tensor_v2(pers.obj(('storage', torch.DoubleStorage, '140245072983168.storage', 'cpu', 8),),
   0,
   (2, 4),
   (4, 1),
   False,
   collections.OrderedDict()),)
```

Both tensors will refer to the tensor in at the path `script_module_v4_unify/constants/140245072983168.storage`.

## Note
According to unify format, all tensors will be written to the folder `.data`, however, torch.jit.load() can't handle the unified format at this moment, so this change will write tensors at the `constants` folders, and mobile will write/read tensors from `constants` folder. such that the model can be interpreted by both jit and mobile.

Differential Revision: [D28375257](https://our.internmc.facebook.com/intern/diff/D28375257/)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge] share tensors in mobile with new api"


As title, the v5 model format will be
```
(base) chenlai@chenlai-mp reuse_constant % zipinfo /Users/chenlai/Documents/pytorch/reuse_constant/tmp/zip/script_module_v5_unify.ptl
Archive:  /Users/chenlai/Documents/pytorch/reuse_constant/tmp/zip/script_module_v5_unify.ptl
Zip file size: 3120 bytes, number of entries: 7
-rw----     0.0 fat       77 bl stor 80-000-00 00:00 script_module_v4_unify/data.pkl
-rw----     0.0 fat      240 bl defN 80-000-00 00:00 script_module_v4_unify/code/__torch__/___torch_mangle_5.py
-rw----     0.0 fat      422 bl defN 80-000-00 00:00 script_module_v4_unify/code/__torch__/___torch_mangle_5.py.debug_pkl
-rw----     0.0 fat       64 bl stor 80-000-00 00:00 script_module_v4_unify/constants/140245072983168.storage
-rw----     0.0 fat      172 bl stor 80-000-00 00:00 script_module_v4_unify/constants.pkl
-rw----     0.0 fat      678 bl stor 80-000-00 00:00 script_module_v4_unify/bytecode.pkl
-rw----     0.0 fat        2 bl stor 80-000-00 00:00 script_module_v4_unify/version
7 files, 1655 bytes uncompressed, 1453 bytes compressed:  12.2%
```
bytecode.pkl is:
```
(5,
 ('__torch__.___torch_mangle_5.TestModule.forward',
  (('instructions',
    (('STOREN', 1, 2),
     ('DROPR', 1, 0),
     ('LOADC', 0, 0),
     ('LOADC', 1, 0),
     ('MOVE', 2, 0),
     ('OP', 0, 0),
     ('LOADC', 1, 0),
     ('OP', 1, 0),
     ('RET', 0, 0))),
   ('operators', (('aten::add', 'int'), ('aten::add', 'Scalar'))),
   ('constants',
    (torch._utils._rebuild_tensor_v2(pers.obj(('storage',
          torch.DoubleStorage,
          '140245072983168.storage',
          'cpu',
          8),),
       0,
       (2, 4),
       (4, 1),
       False,
       collections.OrderedDict()),
     1)),
   ('types', ()),
   ('register_size', 2)),
  (('arguments',
    ((('name', 'self'),
      ('type', '__torch__.___torch_mangle_5.TestModule'),
      ('default_value', None)),
     (('name', 'y'), ('type', 'int'), ('default_value', None)))),
   ('returns',
    ((('name', ''), ('type', 'Tensor'), ('default_value', None)),)))))
```

constants.pkl is:
```
(torch._utils._rebuild_tensor_v2(pers.obj(('storage', torch.DoubleStorage, '140245072983168.storage', 'cpu', 8),),
   0,
   (2, 4),
   (4, 1),
   False,
   collections.OrderedDict()),)
```

Both tensors will refer to the tensor in at the path `script_module_v4_unify/constants/140245072983168.storage`.

## Note
According to unify format, all tensors will be written to the folder `.data`, however, torch.jit.load() can't handle the unified format at this moment, so this change will write tensors at the `constants` folders, and mobile will write/read tensors from `constants` folder. such that the model can be interpreted by both jit and mobile.

Differential Revision: [D28375257](https://our.internmc.facebook.com/intern/diff/D28375257/)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge] share tensors in mobile with new api"


As title, the v5 model format will be
```
(base) chenlai@chenlai-mp reuse_constant % zipinfo /Users/chenlai/Documents/pytorch/reuse_constant/tmp/zip/script_module_v5_unify.ptl
Archive:  /Users/chenlai/Documents/pytorch/reuse_constant/tmp/zip/script_module_v5_unify.ptl
Zip file size: 3120 bytes, number of entries: 7
-rw----     0.0 fat       77 bl stor 80-000-00 00:00 script_module_v4_unify/data.pkl
-rw----     0.0 fat      240 bl defN 80-000-00 00:00 script_module_v4_unify/code/__torch__/___torch_mangle_5.py
-rw----     0.0 fat      422 bl defN 80-000-00 00:00 script_module_v4_unify/code/__torch__/___torch_mangle_5.py.debug_pkl
-rw----     0.0 fat       64 bl stor 80-000-00 00:00 script_module_v4_unify/constants/140245072983168.storage
-rw----     0.0 fat      172 bl stor 80-000-00 00:00 script_module_v4_unify/constants.pkl
-rw----     0.0 fat      678 bl stor 80-000-00 00:00 script_module_v4_unify/bytecode.pkl
-rw----     0.0 fat        2 bl stor 80-000-00 00:00 script_module_v4_unify/version
7 files, 1655 bytes uncompressed, 1453 bytes compressed:  12.2%
```
bytecode.pkl is:
```
(5,
 ('__torch__.___torch_mangle_5.TestModule.forward',
  (('instructions',
    (('STOREN', 1, 2),
     ('DROPR', 1, 0),
     ('LOADC', 0, 0),
     ('LOADC', 1, 0),
     ('MOVE', 2, 0),
     ('OP', 0, 0),
     ('LOADC', 1, 0),
     ('OP', 1, 0),
     ('RET', 0, 0))),
   ('operators', (('aten::add', 'int'), ('aten::add', 'Scalar'))),
   ('constants',
    (torch._utils._rebuild_tensor_v2(pers.obj(('storage',
          torch.DoubleStorage,
          '140245072983168.storage',
          'cpu',
          8),),
       0,
       (2, 4),
       (4, 1),
       False,
       collections.OrderedDict()),
     1)),
   ('types', ()),
   ('register_size', 2)),
  (('arguments',
    ((('name', 'self'),
      ('type', '__torch__.___torch_mangle_5.TestModule'),
      ('default_value', None)),
     (('name', 'y'), ('type', 'int'), ('default_value', None)))),
   ('returns',
    ((('name', ''), ('type', 'Tensor'), ('default_value', None)),)))))
```

constants.pkl is:
```
(torch._utils._rebuild_tensor_v2(pers.obj(('storage', torch.DoubleStorage, '140245072983168.storage', 'cpu', 8),),
   0,
   (2, 4),
   (4, 1),
   False,
   collections.OrderedDict()),)
```

Both tensors will refer to the tensor in at the path `script_module_v4_unify/constants/140245072983168.storage`.

## Note
According to unify format, all tensors will be written to the folder `.data`, however, torch.jit.load() can't handle the unified format at this moment, so this change will write tensors at the `constants` folders, and mobile will write/read tensors from `constants` folder. such that the model can be interpreted by both jit and mobile.

Differential Revision: [D28375257](https://our.internmc.facebook.com/intern/diff/D28375257/)

[ghstack-poisoned]

iseeyuan commented on torch/csrc/jit/mobile/backport_manager.cpp:
```diff
@@ -36,16 +37,47 @@ bool update_bytecode_version(
+
+    // Skip files (exaxt path)
+    for (const auto& excluded_file : excluded_files) {
+      if (record == excluded_file) {
```
Nice.

iseeyuan reviewed (approved):
LGTM. Thanks!

raziel reviewed (approved):
Thanks a lot @cccclai
